crates/file/src/blob.rs

```diff
@@ -0,0 +1,338 @@
+/// types to be used to create a `Blob`. Ignore this, and use &[u8], &str, etc to create a `Blob`.
+///
+/// The trait is sealed: it must only be implemented by types in this
+/// crate, as this crate relies on invariants regarding the `JsValue` returned from `into_jsvalue`.
+pub trait BlobContents: Sealed {
+    fn into_jsvalue(self) -> JsValue;
+}
+
+impl<'a> Sealed for &'a str {}
+impl<'a> BlobContents for &'a str {
+    fn into_jsvalue(self) -> JsValue {
+        JsValue::from_str(self)
+    }
+}
+
+impl<'a> Sealed for &'a [u8] {}
+impl<'a> BlobContents for &'a [u8] {
+    fn into_jsvalue(self) -> JsValue {
+        let array = unsafe { js_sys::Uint8Array::view(self) };
+        js_sys::Uint8Array::new(&array).into()
+    }
+}
+
+impl Sealed for js_sys::ArrayBuffer {}
+impl BlobContents for js_sys::ArrayBuffer {
+    fn into_jsvalue(self) -> JsValue {
+        self.into()
+    }
+}
+
+/// A [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob).
+///
+/// `Blob`s can be created directly from `&str`, `&[u8]`, and `js_sys::ArrayBuffer`s using the
+/// `From::from` implementation (e.g. `let b: Blob = "test".into()`), or using the
+/// `new_with_options` constructor.
+#[derive(Debug, Clone)]
+pub struct Blob {
+    inner: web_sys::Blob,
+}
+
+impl Blob {
+    fn new(val: JsValue, mime_type: Option<&str>) -> Blob {
+        let mut properties = web_sys::BlobPropertyBag::new();
+        if let Some(mime_type) = mime_type {
+            properties.type_(mime_type);
+        }
+
+        let parts = js_sys::Array::of1(&val);
+        let inner = web_sys::Blob::new_with_u8_array_sequence_and_options(&parts, &properties);
+
+        Blob::from(inner.unwrap_throw())
+    }
+
+    /// Like `new`, but allows specifying the media type (also known as *content type* or *mime
+    /// type*) of the `Blob`.
+    pub fn new_with_options<T>(content: T, mime_type: Option<&str>) -> Blob
+    where
+        T: BlobContents,
+    {
+        Blob::new(content.into_jsvalue(), mime_type)
+    }
+
+    pub fn slice(&self, start: u64, end: u64) -> Self {
+        let start = safe_u64_to_f64(start);
+        let end = safe_u64_to_f64(end);
+
+        let b: &web_sys::Blob = self.as_ref();
+        Blob::from(b.slice_with_f64_and_f64(start, end).unwrap_throw())
+    }
+
+    /// The number of bytes in the Blob/File.
+    pub fn size(&self) -> u64 {
+        safe_f64_to_u64(self.inner.size())
+    }
+
+    /// The typed *media type* (also known as *content type* or *mime type*) of the `File` or
+    /// `Blob`.
+    #[cfg(feature = "mime")]
+    pub fn mime_type(&self) -> Result<mime::Mime, mime::FromStrError> {
+        self.raw_mime_type().parse()
+    }
+
+    /// The *media type* (also known as *content type* or *mime type*) of the `File` or
+    /// `Blob` as a `String`.
+    pub fn raw_mime_type(&self) -> String {
+        self.inner.type_()
+    }
+}
+
+impl<'a> From<&'a str> for Blob {
```

Comment:
I'm not so sure about these impls, I think they're covered by `Blob::new`.

Reply:
Yes they're just there because they can be (i.e. there is a sensible conversion). I can remove them if you like?

Reply:
I'd rather us not have them. We can always add them later, but we can't remove them after we've added them.